File: ejerc.diccionarios/Ejercicio 10/Ejercicio10.java
// Importa las librerías necesarias
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class GestionClientes {
    public static void main(String[] args) {
        // Diccionario principal donde se almacenan los clientes
        // La clave es el NIF (String) y el valor es otro Map con los datos del cliente
        Map<String, Map<String, Object>> clientes = new HashMap<>();
        Scanner sc = new Scanner(System.in); // Objeto para leer datos por consola

        // Bucle infinito para mostrar el menú hasta que el usuario decida salir
        while (true) {
            // Menú de opciones
            System.out.println("\nMenú de opciones:");
            System.out.println("1. Añadir cliente");
            System.out.println("2. Eliminar cliente");
            System.out.println("3. Mostrar cliente");
            System.out.println("4. Listar todos los clientes");
            System.out.println("5. Listar clientes preferentes");
            System.out.println("6. Terminar");
            System.out.print("Seleccione una opción: ");
            String opcion = sc.nextLine().trim(); // Lee la opción elegida

            // Estructura switch para manejar cada opción del menú
            switch (opcion) {
                case "1": // Opción para añadir un cliente
                    System.out.print("Ingrese el NIF del cliente: ");
                    String nif = sc.nextLine();

                    // Se crea un diccionario para guardar los datos del cliente
                    Map<String, Object> datos = new HashMap<>();

                    // Se piden los datos al usuario
                    System.out.print("Ingrese el nombre: ");
                    datos.put("nombre", sc.nextLine());

                    System.out.print("Ingrese la dirección: ");
                    datos.put("direccion", sc.nextLine());

                    System.out.print("Ingrese el teléfono: ");
                    datos.put("telefono", sc.nextLine());

                    System.out.print("Ingrese el correo: ");
                    datos.put("correo", sc.nextLine());

                    System.out.print("¿Es cliente preferente? (s/n): ");
                    datos.put("preferente", sc.nextLine().trim().equalsIgnoreCase("s"));

                    // Se agrega el cliente al mapa principal usando el NIF como clave
                    clientes.put(nif, datos);
                    System.out.println("Cliente añadido con éxito.");
                    break;

                case "2": // Opción para eliminar cliente
                    System.out.print("Ingrese el NIF del cliente a eliminar: ");
                    String nifEliminar = sc.nextLine();

                    // Se intenta eliminar el cliente. Si existe, se elimina; si no, se notifica
                    if (clientes.remove(nifEliminar) != null) {
                        System.out.println("Cliente eliminado.");
                    } else {
                        System.out.println("Cliente no encontrado.");
                    }
                    break;

                case "3": // Opción para mostrar datos de un cliente específico
                    System.out.print("Ingrese el NIF del cliente a mostrar: ");
                    String nifMostrar = sc.nextLine();

                    // Verifica si el cliente existe y muestra su información
                    if (clientes.containsKey(nifMostrar)) {
                        System.out.println("Datos del cliente: " + clientes.get(nifMostrar));
                    } else {
                        System.out.println("Cliente no encontrado.");
                    }
                    break;

                case "4": // Opción para listar todos los clientes
                    System.out.println("\nLista de todos los clientes:");
                    for (String key : clientes.keySet()) {
                        System.out.println("NIF: " + key + ", Nombre: " + clientes.get(key).get("nombre"));
                    }
                    break;

                case "5": // Opción para listar únicamente clientes preferentes
                    System.out.println("\nClientes preferentes:");
                    for (Map.Entry<String, Map<String, Object>> entry : clientes.entrySet()) {
                        // Verifica si el cliente tiene el atributo "preferente" en verdadero
                        if ((boolean) entry.getValue().get("preferente")) {
                            System.out.println("NIF: " + entry.getKey() + ", Nombre: " + entry.getValue().get("nombre"));
                        }
                    }
                    break;

                case "6": // Terminar programa
                    System.out.println("Programa terminado.");
                    sc.close(); // Cierra el Scanner para liberar recursos
                    return; // Finaliza el método main

                default: // Opción no válida
                    System.out.println("Opción inválida.");
                    break;
            }
        }
    }
}

File: ejerc.diccionarios/Ejercicio 10/Ejercicio10.py
# Diccionario para almacenar los clientes
# La clave será el NIF (identificación única), y el valor será otro diccionario con los datos del cliente.
clientes = {}

# Bucle infinito para mostrar el menú hasta que el usuario elija terminar
while True:
    # Menú principal
    print("\nMenú de opciones:")
    print("1. Añadir cliente")
    print("2. Eliminar cliente")
    print("3. Mostrar cliente")
    print("4. Listar todos los clientes")
    print("5. Listar clientes preferentes")
    print("6. Terminar")
    
    # Solicitamos al usuario una opción del menú
    opcion = input("Seleccione una opción: ").strip()

    # --- OPCIÓN 1: Añadir cliente ---
    if opcion == "1":
        # Se solicita al usuario la información del cliente
        nif = input("Ingrese el NIF del cliente: ").strip()
        nombre = input("Ingrese el nombre: ").strip()
        direccion = input("Ingrese la dirección: ").strip()
        telefono = input("Ingrese el teléfono: ").strip()
        correo = input("Ingrese el correo: ").strip()
        # Se pregunta si es preferente; si el usuario escribe 's', se guarda como True
        preferente = input("¿Es cliente preferente? (s/n): ").strip().lower() == 's'
        
        # Se guarda el cliente en el diccionario, con un subdiccionario que contiene todos sus datos
        clientes[nif] = {
            "nombre": nombre,
            "direccion": direccion,
            "telefono": telefono,
            "correo": correo,
            "preferente": preferente
        }
        print("Cliente añadido con éxito.")

    # --- OPCIÓN 2: Eliminar cliente ---
    elif opcion == "2":
        # Se pide el NIF del cliente a eliminar
        nif = input("Ingrese el NIF del cliente a eliminar: ").strip()
        # Se verifica si existe el cliente
        if nif in clientes:
            del clientes[nif]  # Eliminamos el cliente del diccionario
            print("Cliente eliminado.")
        else:
            print("Cliente no encontrado.")

    # --- OPCIÓN 3: Mostrar datos de un cliente específico ---
    elif opcion == "3":
        nif = input("Ingrese el NIF del cliente a mostrar: ").strip()
        if nif in clientes:
            # Se muestra el diccionario con los datos del cliente
            print("Datos del cliente:", clientes[nif])
        else:
            print("Cliente no encontrado.")

    # --- OPCIÓN 4: Listar todos los clientes ---
    elif opcion == "4":
        print("\nLista de todos los clientes:")
        # Se recorre el diccionario mostrando el NIF y el nombre de cada cliente
        for nif, datos in clientes.items():
            print(f"NIF: {nif}, Nombre: {datos['nombre']}")

    # --- OPCIÓN 5: Listar clientes preferentes ---
    elif opcion == "5":
        print("\nClientes preferentes:")
        # Se recorre el diccionario filtrando solo aquellos que son preferentes
        for nif, datos in clientes.items():
            if datos['preferente']:
                print(f"NIF: {nif}, Nombre: {datos['nombre']}")

    # --- OPCIÓN 6: Salir del programa ---
    elif opcion == "6":
        print("Programa terminado.")
        break  # Rompe el bucle y finaliza el programa

    # --- OPCIÓN NO VÁLIDA ---
    else:
        print("Opción inválida.")
